arangod/Graph/GraphOperations.cpp

```diff
@@ -605,22 +589,122 @@ OperationResult GraphOperations::updateEdge(const std::string& definitionName,
+
   return modifyDocument(definitionName, key, document, true, std::move(rev),
-                        waitForSync, returnOld, returnNew, keepNull);
+                        waitForSync, returnOld, returnNew, keepNull, trx);
 }
 
 OperationResult GraphOperations::replaceEdge(const std::string& definitionName,
                                              const std::string& key, VPackSlice document,
                                              boost::optional<TRI_voc_rid_t> rev,
                                              bool waitForSync, bool returnOld,
                                              bool returnNew, bool keepNull) {
+  std::string fromCollectionName;
+  std::string fromCollectionKey;
+  std::string toCollectionName;
+  std::string toCollectionKey;
+
+  OperationResult res = validateEdgeContent(document, fromCollectionName, fromCollectionKey,
+                                            toCollectionName, toCollectionKey);
+
+  if (res.fail()) {
+    return res;
+  }
+
+  std::vector<std::string> readCollections;
+  std::vector<std::string> writeCollections;
+
+  readCollections.emplace_back(fromCollectionName);
+  readCollections.emplace_back(toCollectionName);
+  writeCollections.emplace_back(definitionName);
+
+  transaction::Options trxOptions;
+  trxOptions.waitForSync = waitForSync;
+  transaction::Methods trx(ctx(), readCollections, writeCollections, {}, trxOptions);
+
+  Result tRes = trx.begin();
+
+  if (!tRes.ok()) {
+    return OperationResult(tRes);
+  }
+
+  res = validateEdgeVertices(fromCollectionName, fromCollectionKey,
+                             toCollectionName, toCollectionKey, trx);
+
+  if (res.fail()) {
+    return res;
+  }
+
   return modifyDocument(definitionName, key, document, false, std::move(rev),
-                        waitForSync, returnOld, returnNew, keepNull);
+                        waitForSync, returnOld, returnNew, keepNull, trx);
 }
 
-OperationResult GraphOperations::createEdge(const std::string& definitionName,
-                                            VPackSlice document,
-                                            bool waitForSync, bool returnNew) {
+OperationResult GraphOperations::validateEdgeVertices(
+    const std::string& fromCollectionName, const std::string& fromCollectionKey,
+    const std::string& toCollectionName, const std::string& toCollectionKey,
+    transaction::Methods& trx) {
+  VPackBuilder bT;
+  {
+    VPackObjectBuilder guard(&bT);
+    bT.add(StaticStrings::KeyString, VPackValue(toCollectionKey));
+  }
+
+  VPackBuilder bF;
+  {
+    VPackObjectBuilder guard(&bF);
+    bF.add(StaticStrings::KeyString, VPackValue(fromCollectionKey));
+  }
+
+  OperationOptions options;
+  OperationResult resultFrom = trx.document(fromCollectionName, bF.slice(), options);
+  OperationResult resultTo = trx.document(toCollectionName, bT.slice(), options);
+
+  if (!resultFrom.ok() || !resultTo.ok()) {
+    // actual result doesn't matter here
+    trx.finish(Result());
```

Comment:
This will commit the transaction, i guess in the current code this will work (we have only read upto now) but to ba save against refactorings please make sure to ABORT the trx here. (by using the invalid Result as input)

Reply:
done, please verify and resolve.